Handle hardlinks during extraction.

bruce-one opened the pull request:
The runtime extract current code doesn't handle hardlinks explicitly and will extract a hardlinked file multiple times.

This PR copies the unsquashfs logic to do so.

(The code itself might be terrible :-s but it was mostly just a transplant so hopefully it's not too bad....)

(My understanding is that we're in a world of hurt if the malloc on 383 failed? But... I wasn't sure whether to forcibly exit or not...)

probonopd commented:
Thanks @bruce-one. Since you seem knowledgable about those things, could you have a a look at our `inode.base.inode_type` 9 and 10 issue as well? (What are they, after all?) 

* https://github.com/AppImage/AppImageKit/issues/498
* https://github.com/AppImage/appimaged/issues/13
* https://github.com/AppImage/AppImageKit/issues/279 (even more strange inode types)

Thanks again.

probonopd commented:
Do you have a test AppImage that contains hardlinks, and how can this be tested?

bruce-one commented:
I'll put together a test :-) (I basically just created a dir with some hard links (eg `mkdir blah.AppDir; touch blah.AppDir/a; ln blah.AppDir/a blah.AppDir/b`; made the app image and then extracted it; and then checked that the two extracted files had the same inode (eg via `ls -i blah.AppDir`) -- but I'll have a look at the test suite and put something together :-) )

I'll be offline for a couple of days, but I'll get it done :-)

I'll have a look at those other issues and see if I can figure anything out :-)

TheAssassin commented:
Looks good to me code wise. I can't tell whether this is solid functionality wise, but it looks good to me so far.

TheAssassin commented:
Adding this PR to my TODO list for next week.

TheAssassin commented on src/runtime.c:
```diff
@@ -404,6 +412,16 @@ main (int argc, char *argv[])
                             }
                         }
                     } else if(inode.base.inode_type == SQUASHFS_REG_TYPE || inode.base.inode_type == SQUASHFS_LREG_TYPE){
+                        if(created_inode[inode.base.inode_number - 1]) {
```
I'm not entirely sure why this check is performed. Would you please add a comment or two explaining what you do?

TheAssassin commented on src/runtime.c:
```diff
@@ -412,15 +413,18 @@ main (int argc, char *argv[])
-                        if(created_inode[inode.base.inode_number - 1]) {
+                        // if we've already created this inode, then this is a hardlink
+                        char* exiting_path_for_inode = created_inode[inode.base.inode_number - 1];
+                        if(exiting_path_for_inode) {
```
Is this supposed to be a `NULL` check? Please try to use explicit comparisons, e.g., `exiting_path_for_inode == NULL`.

TheAssassin commented:
Looks good now. I'll try to review, test and merge this PR tonight. Thanks @bruce-one.

TheAssassin commented on src/runtime.c:
```diff
@@ -379,6 +380,14 @@ main (int argc, char *argv[])
+        if(created_inode != NULL) {
+            memset(created_inode, 0, fs.sb.inodes * sizeof(char *));
+        } else {
+            fprintf(stderr, "Can't track hardlinks.\n");
```
I just saw your question. Yes, if that isn't possible, I'd rather exit instead of falling back to creating copies. This way, the users can report the issues to us, and we can hopefully fix them. I don't like implicit fallbacks.

bruce-one commented:
[example.tar.gz](https://github.com/AppImage/AppImageKit/files/2255601/example.tar.gz)
$ ./Hardlink_test-x86_64.broken.AppImage  ; echo ---; ./Hardlink_test-x86_64.fixed.AppImage ; echo ---;  rm -rf squashfs-root/; ./Hardlink_test-x86_64.broken.AppImage --appimage-extract ; ./squashfs-root/AppRun; echo ---; rm -rf squashfs-root/; ./Hardlink_test-x86_64.fixed.AppImage --appimage-extract ; ./squashfs-root/AppRun

128 find
128 find.link
128 find.link2
128 find.link3
192 script.sh
---
128 find
128 find.link
128 find.link2
128 find.link3
192 script.sh
---
squashfs-root/.DirIcon
squashfs-root/AppRun
squashfs-root/blah.desktop
squashfs-root/blah.png
squashfs-root/usr
squashfs-root/usr/bin
squashfs-root/usr/bin/find
squashfs-root/usr/bin/find.link
squashfs-root/usr/bin/find.link2
squashfs-root/usr/bin/find.link3
squashfs-root/usr/bin/script.sh
28180234 find
28180235 find.link
28180236 find.link2
28180237 find.link3
28180238 script.sh
---
squashfs-root/.DirIcon
squashfs-root/AppRun
squashfs-root/blah.desktop
squashfs-root/blah.png
squashfs-root/usr
squashfs-root/usr/bin
squashfs-root/usr/bin/find
squashfs-root/usr/bin/find.link
squashfs-root/usr/bin/find.link2
squashfs-root/usr/bin/find.link3
squashfs-root/usr/bin/script.sh
28180246 find
28180246 find.link
28180246 find.link2
28180246 find.link3
28180247 script.sh

TheAssassin commented:
Thanks @bruce-one!

bruce-one commented:
No problem, thank you and @probonopd for everything you do with appimage!
